base/broadcast.jl

```diff
@@ -333,13 +335,32 @@ end
 end
 @inline broadcast_c(f, ::Type{Any}, a...) = f(a...)
 @inline broadcast_c(f, ::Type{Tuple}, A, Bs...) =
-    tuplebroadcast(f, first_tuple(A, Bs...), A, Bs...)
+    tuplebroadcast(f, tuplebroadcast_maxtuple(A, Bs...), A, Bs...)
 @inline tuplebroadcast(f, ::NTuple{N,Any}, As...) where {N} =
     ntuple(k -> f(tuplebroadcast_getargs(As, k)...), Val(N))
 @inline tuplebroadcast(f, ::NTuple{N,Any}, ::Type{T}, As...) where {N,T} =
     ntuple(k -> f(T, tuplebroadcast_getargs(As, k)...), Val(N))
-first_tuple(A::Tuple, Bs...) = A
-@inline first_tuple(A, Bs...) = first_tuple(Bs...)
+# When the result of broadcast is a tuple it can only come from mixing n-tuples
+# of the same length with scalars and 1-tuples. So, in order to have a
+# type-stable broadcast, we need to find a tuple of maximum length (except when
+# there are only scalars, empty tuples and 1-tuples, in which case the
+# returned value will be an empty tuple).
+# The following methods compare broadcast arguments pairwise to determine the
+# length of the final tuple.
+tuplebroadcast_maxtuple(A, B) =
+    _tuplebroadcast_maxtuple(containertype(A), containertype(B), A, B)
+@inline tuplebroadcast_maxtuple(A, Bs...) =
+    tuplebroadcast_maxtuple(A, tuplebroadcast_maxtuple(Bs...))
+tuplebroadcast_maxtuple(A::NTuple{N,Any}, ::NTuple{N,Any}...) where {N} = A
```

Comment:
I think this needs to be a `_tuplebroadcast_maxtuple` method, for example with the testcase `broadcast(+, (1, 2), (3, 4), 5)`.

Reply:
That example works with this PR as is. I had it like this to reduce the compiler/inference work in cases where you have a bunch of tuples of the same length and a scalar element (which I imagined would be common), e.g. `(1,2) .+ (2,3) .+ (3,4) .+ (4,5) .+ 1`. But I can make it a `_tuplebroadcast_maxtuple` for conciseness if preferred.

Reply:
Sorry, I screwed up the order of my example. It should be `broadcast(+, 5, (1, 2), (3, 4))` which I believe would not currently work.

Reply:
I can confirm that example (or `broadcast(+, 5, (1, 2), (3, 4))`) works fine.

`tuplebroadcast_maxtuple` will call `_tuplebroadcast_maxtuple` only when the number or arguments to `tuplebroadcast_maxtuple` is two and both are not tuples of the same lengths (if the they are dispatch would select `tuplebroadcast_maxtuple(A::NTuple{N,Any}, ::NTuple{N,Any}...) where {N}` instead).

Reply:
Oh, I understand. Thanks for the correction.